packages/connect-node/src/http2-session-manager.ts

```diff
@@ -653,23 +686,40 @@ function ready(
 
   function onIdleTimeout() {
     conn.close();
-    onClose(); // trigger a state change right away so we are not open to races
+    onClose(); // trigger a state change right away, so we are not open to races
   }
 
   function onGoaway(
     errorCode: number,
     lastStreamID: number,
-    opaqueData: Buffer
+    opaqueData: Buffer | undefined | null
   ) {
     const tooManyPingsAscii = Buffer.from("too_many_pings", "ascii");
     if (
       errorCode === http2.constants.NGHTTP2_ENHANCE_YOUR_CALM &&
+      opaqueData != null &&
       opaqueData.equals(tooManyPingsAscii)
     ) {
       // double pingIntervalMs, following the last paragraph of https://github.com/grpc/proposal/blob/0ba0c1905050525f9b0aee46f3f23c8e1e515489/A8-client-side-keepalive.md#basic-keepalive
       options.pingIntervalMs = options.pingIntervalMs * 2;
       receivedGoAwayEnhanceYourCalmTooManyPings = true;
     }
+    if (errorCode === http2.constants.NGHTTP2_NO_ERROR) {
+      receivedGoAwayNoError = true;
+      const nodeMajor = parseInt(process.versions.node.split(".")[0], 10);
+      // Node.js v16 closes a connection on its own when it receives a GOAWAY
+      // frame and there are no open streams (emitting a "close" event and
+      // destroying the session), but more recent versions do not.
+      if (streamCount == 0 && nodeMajor >= 18) {
+        conn.destroy(
+          new ConnectError(
+            "received GOAWAY without any open streams",
+            Code.Canceled
+          ),
+          http2.constants.NGHTTP2_NO_ERROR
+        );
+      }
+    }
```

Comment:
In case it's relevant:
- grpc.js seems to always close the connection https://github.com/grpc/grpc-node/blob/186718cb70af708c75584909b21169289d7363b2/packages/grpc-js/src/transport.ts#L198-L212
- the server will also do the same, close the connection after sending `goaway` https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md#server-enforcement

Reply:
>     // Node.js v16 closes a connection on its own when it receives a GOAWAY
>     // frame and there are no open streams (emitting a "close" event and
>     // destroying the session), but more recent versions do not.

Does it mean the bug shouldn't happen in node 16? I'm using node 16 and I can confirm it does happen

Reply:
I'm not sure if `@grpc/grpc-js` is really _always_ closing the connection when receiving a GOAWAY. I've updated the PR description with details about the graceful shutdown. I would expect an HTTP/2 client to drain streams first. It is possible that `@grpc/grpc-js` does not do so for historical reasons.

Note that there is a distinction between error code NO_ERROR and actual error codes, and also the special behavior in case of ENHANCE_YOUR_CALM (last paragraph in [basic keepalive](https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md#basic-keepalive)).

Regarding the comment: The `http2` module implements graceful shutdown. It rejects new streams, and it automatically closes the connection after all streams have finished. But if a connection _without_ open streams receives a GOAWAY with NO_ERROR, behavior differs between versions. Seeing https://github.com/bufbuild/connect-es/issues/680 in v16 is expected.